### ISSUE TITLE
[RFC] Move tests out of the source and source out of the tests

### ISSUE THREAD

**Issue by iltar** [symfony/symfony#17749]
#### The Source
- Often when I'm auto-completing classes in my IDE, I get the suggestion to use the *Test file.
- Project wide searches and Usage Finder always shows tests using the search key or method/class etc.
- KernelTestCase and WebTestCase are library dependencies in a test directory

I'm working on adding some functional tests to a bundle and this gave me a problem. I had to depend on the FrameworkBundle 2.3 for BC reasons but I couldn't depend on 2.3 because that one didn't have the KernelTestCase yet so I needed 2.5.
#### The Symptoms

For applications [the recommended location is in /tests/](http://symfony.com/doc/current/quick_tour/the_architecture.html#understanding-the-directory-structure). For Application bundles [this seem to be /Tests/](http://symfony.com/doc/current/best_practices/tests.html) and [the same goes for re-usable bundles](http://symfony.com/doc/current/cookbook/bundles/best_practices.html#directory-structure).

The current setup with the recommendations and implementations cause a few issues:
- I need to load files from the vendor Tests in order to run my tests.
- I get vendor test files which I can autoload and depend on.
- 2 different recommendations for test locations.
- I cannot easily exclude them from the distributed version.
#### The Solution

Personally I always put my code in bundles in the /src/ directory, this gives me a clean root directory and my tests in /test/. This leaves the root directory for all the meta-information for your package ([example](https://github.com/iltar/http-bundle)). So my suggestion is:
- Extract the `WebTestCase` and `KernelTestCase` to another component (PhpunitBridge?) where they can reside in the source instead of test location.
- Exclude the tests from the [archive](https://getcomposer.org/doc/04-schema.md#archive).
- ~~Not put the source in the package root but in /src/ like suggested for applications.~~ Not going to happen due to subtree split.
- Put tests in /test/ or /tests/ where they have an autoload-dev so they cannot be loaded in production code.

This will create a clearer structure where your source is not polluted by the tests and you cannot put anything of your public api in your tests (like the TestCases). When the test-cases are moved out, I can also require 2.3 or higher instead of 2.5 and higher (thus 2.7).

**Comment by dosten** [comment]
This topic has been discussed some time ago, the major problem with this idea is the subtree split of each component.

**Comment by iltar** [comment]
@dosten Yeah I got informed about that. I've updated the issue.

**Comment by jakzal** [comment]
We used to have tests outside of the src directory (see https://github.com/symfony/symfony/commit/fea6b79acdf01b79ac3d3206c001a9dec662da93).

We usually put test cases outside of the tests namespace already, so they're not part of the `Tests` namespace. Both WebTestCase and KernelTestCase are already in the `Symfony\Bundle\FrameworkBundle\Test` namespace (the first one was moved recently).

**Comment by sstok** [comment]
To me there is a difference between tests `tests/` (internal, should not be extended in other packages) and test helpers `test/` (traits/test-cases). When you "need" (there is no other way) to extend a test helper that is in `tests/` it should be moved to `test/` so you can use it.

I actually have that problem with some Doctrine tests, 
https://github.com/rollerworks/search-doctrine-orm/blob/master/tests/OrmTestCase.php

Which I solved using https://github.com/rollerworks/search-doctrine-orm/blob/master/composer.json#L36

But there is a risk of breakage here as I'm using classes that are not part of the public API and therefor have no BC policy.

> Often when I'm auto-completing classes in my IDE, I get the suggestion to use the *Test file.

Mark tests directories as excluded :+1: *going to that now as I to have fallen trap to this...*

**Comment by iltar** [comment]
@sstok, that's 68 (symfony: 49) directories in total which I'd rather not even have on my filesystem to begin with. I only need them when I explicitly install that repository and I have to run them. Given I'd exclude Tests via a regex, the kernel/web tests are excluded too. The solution is valid but won't work if you have 20 applications and dozens of bundles which all use symfony.

**Comment by Tobion** [comment]
This was also done with gitattributes and reverted in #6605
I still think this should be reintroduced as I said in https://github.com/symfony/symfony/issues/6605#issuecomment-60306001

**Comment by robfrawley** [comment]
- Tests are documentation and should remain next to their src.
- For production releases, I assume everyone has their own prefered deploy mechanisms: removing files not needed for such an env can easily be accomplished during your preferred deploy operation.
- I too use PHPStorm: I don't understand the difficulty of setting the tests directory to ignored within the IDE if such is desired.

**Comment by iltar** [comment]
@robfrawley Because if you use symfony/symfony and have 20+ applications doing this, you'll spend too much time ignoring them, where as excluding tests from the packaged version is something you do with your own package, not that of others.

**Comment by nicolas-grekas** [comment]
An other option: https://github.com/editorconfig/editorconfig/issues/228

**Comment by nicolas-grekas** [comment]
I also just sent this message to github support:
> ZIP files downloaded from github honor the export-ignore attributes set in .gitattributes
Is there a way to get a ZIP that includes ALL files (i.e. that is built without taking .gitattributes into account?)
That would be really great because that would allow us to have packages for our project: one for humans (with the full source) and one for robots/ci-scripts (the current one).

If anyone has the answer (or could push for making this happen), help welcomed :)

**Comment by nicolas-grekas** [comment]
An other option would be to add a feature to composer to clean paths specified in the `archive.exclude` option in composer.json files.

**Comment by Tobion** [comment]
One more reasons to exclude tests:

Composer beta
> Added possibility to configure per-package preferred-install types for more flexibility between prefer-source and prefer-dist

**Comment by nicolas-grekas** [comment]
Here is my reasoning pro keeping tests where they are and also in the .zip
archives:

1. Some of us want the tests out, either because e.g "they pollute their
autocompleter, or their code searches or just their filesystem".
2. Some of us want the tests in, because e.g. "having them always at hand
enables this coding workflow of looking at tests to know how the code
behaves and how it should be used".
3. You can grow the list of pro and contra arguments, the truth is that
everyone is right: this is a matter of personal preference.
4. In this situation, we, as the Symfony project, can have two attitudes:
we can forcibly decide for others, or we can empower people in allowing
both way of things.
5. Given the history of this very topic, and the presented arguments at
both sides, I'm now convinced that we should not enforce what would be
otherwise a subjective and controversial "best way".
6. So the question for me is now: which choice allows both practices to
coexist?
7. The first aspect is the if we always ship the tests as we do today, 2.
is allowed, and 1. is just one IDE configuration or cleaning script away.
8. Moving tests out of each components folder breaks 2. for subtree splits.
Looks like a no-go.
9. Now moving tests out of archives: since we install code through
composer, we could use --prefer-dist for 1. and --prefer-source for 2. But
this ignores one practical and fundamental difference: --prefer-source is
slow as hell. --prefer-dist has one ZIP to download, and then has local
caching, which means it's fast, very fast. --prefer-source has to `git
clone` and isn't cached as of now. Nobody can't possibly accept to loose so
many time because "hey, 2. is your choice, haha". So, my point is that in
practice, this has the same effect as not shipping the tests at all. No-go
again.

In conclusion: I'm all for the status quo, which is the best solution,
*unless* --prefer-source can be made as fast as --prefer-dist (or any other
way that would make fetching the tests as fast, see e.g. my email to github
support). Meanwhile, I'm all for enhancing editorconfig or any other way to
help 1. supporters configure their IDE or clean their filesystem.

**Comment by iltar** [comment]
How many people actually spend time diving into tests and trying to understand them (which is usually more complicated than looking at the code)? If people really want to see the tests while they are developing they can choose already: `--prefer-source`. If this is too slow for them, they can also view the files on github.

By packaging the tests, people who don't use them or don't need them (My guess is 90%+) still get them. This is extra bandwidth, extra actions during deployment and extra effort in the IDE. If people really, really want to run the tests, they should make a clone (or grab the zip from github) and run it in there with their `require-dev` dependencies.

When the package comes from Packagist it should not include test files for the simple reason that most people won't even need them and should follow one of the above steps for a better workflow in the first place.

In my eyes this would mean:
 - Leaving the repository as is for github
 - Adding composer excludes for packaging

**Comment by Slamdunk** [comment]
> You can grow the list of pro and contra arguments, the truth is that everyone is right: this is a matter of personal preference.

@nicolas-grekas no, it's not a matter of personal preference. There are situations where you can do everything, but it's better to do only one thing (like putting your sources outside the `./public` folder).

There are two environments with different needs, dev & prod, but we have to face the problems in the order of priority, so let's discuss prod before dev.
In production we have two topics: security and performance.
**NOT** having additional files will *always* improve both.
Security: the only code that can't be broken, or can't broke something else, is the one that there isn't.
Performance: less code, less resources, less timings.
The `composer.json > autoload > exclude-from-classmap` already present it's good, but not enough.
I am sad that you closed your https://github.com/symfony/symfony/pull/17995 about `.gitattributes`: I highly hope that to be merged.

Once the production is faced, only then we can discuss about development, and whether or not including additional informations (like tests) in a dev setup.

**Comment by Slamdunk** [comment]
@nicolas-grekas any consideration about my previous thought and reopening https://github.com/symfony/symfony/pull/17995?

**Comment by nicolas-grekas** [comment]
No way for me: I won't take the lead on this topic...

**Comment by mlocati** [comment]
What about this solution for the problem? https://github.com/composer/composer/issues/5367

**Comment by yakobe** [comment]
Just out of interest, why does symfony [recommend that the tests are in the root of a project](http://symfony.com/doc/current/best_practices/creating-the-project.html#structuring-the-application) but do not do this in symfony/symfony? I understand that there would be issues with the subtree split, but i’m sure something could be worked out?

I’m not trying to force a change here but just understand the reasoning so I can work out my own best practice. Having the tests together in root seems like a logical and friendly solution. However symfony moved them out, does this mean the more complex projects get I would probably need to move them out too?

**Comment by stof** [comment]
@yakobe because we want to include tests in our subtree split repositories, and for that, they need to be in the split folder.

**Comment by yakobe** [comment]
@stof yes, as I say, I understand about that. However, as far as i recall this is some automated process (i think with split.sh) so there is probably a way around this to make it work.

It just seems very strange to say "We recommend you do it like this, but we do it differently". If symfony/symfony immediately hit an unsolveable problem with this structure, perhads there is something inherntly wrong?

**Comment by nicolas-grekas** [comment]
@yakobe if you read official Symfony best practices, you may notice that there are different for non-reusable apps (ie your private project) than for shared packages/bundles. The recos that applies to apps are made to prevent doing things that bundle require, but are just overengineering to private apps.
The same apply here. symfony/symfony is structured to reach its goals, which is not the goal of your private apps.

**Comment by nicolas-grekas** [comment]
@iltar can we close this one? From the symfony project pov, this is solved to me, it's an argumented "no". I understand you don't agree, but that's still "resolved".

**Comment by iltar** [comment]
@nicolas-grekas sure!

**Comment by yakobe** [comment]
@nicolas-grekas thanks, yes I understand that. Lots of private apps sometimes end up public / reusable tho. If this structure locks you in, or may create problems down the line, then it might not be advisable as a general standard. If there is a workaround to deal with it, mayne it would be good to standardise symfony/symfony too?

Please dont think i am trolling you or trying to cause an problem here 😄. That is really not my intention and I am very grateful for all the hard work and though you guys put in 👍. I am just attempting to gain understanding and maybe offer perspective. Perhaps the solution is a mutually agreeable one? Perhaps there is a solution but "it just isn't worth it". Resistance to change without consideration can lead to unfortunate situations in the future.

Perhaps this issue is just not high prio or maybe the discussion should be taken elsewhere?

**Comment by nicolas-grekas** [comment]
symfony/symfony a meta project: the project levels are the subtree splits. At the project level, tests are at the root. That' what @stof mentioned also.

**Comment by iltar** [comment]
I think this issue of symfony/symfony root vs package root will become a lot less once flex is released.

**Comment by yakobe** [comment]
@nicolas-grekas that makes it clearer to see it as a "meta" project. A bit strange that work is done on the meta project but i guess that is thats a side affect of the monolithic project approach. Definitely makes sense in this case tho, so thank you.

@iltar could you point me in the right direction for info on flex?

**Comment by javiereguiluz** [comment]
@yakobe there's no public info on "Symfony Flex" project yet. It will be released "soon" and it will be a new way to install/manage Symfony applications in a "composable" way.

**Comment by stof** [comment]
@yakobe This was only announced during the SymfonyCon keynote for now, and the video is not yet available AFAIK.

**Comment by yakobe** [comment]
Great, I will keep a look out for it. Thanks for the info everyone 👍
